src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Protocol/AutoInsert/RemoteInsertTextEdit.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.Razor.AutoInsert;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.Protocol.AutoInsert;
+
+// TODO: check annotations
+[DataContract]
+internal readonly record struct RemoteInsertTextEdit(
+        [property: DataMember(Name = "textEdit")]
+        TextEdit TextEdit,
```

Comment:
Sadly, this is not going to work as it is right now.

We have two options for remote comms, MessagePack or Json. If using MessagePack then we need `[DataContract]`, which you have, and `[DataMember]` on each property, but the data member attributes must specify the `Order` parameter. `TextEdit` is not attributed in that way, so won't work with message pack.

Your options are Json serialization, or using Roslyn's `TextChange` instead of `TextEdit` which is attributed for message pack. Given a choice, I'd go with the latter since Auto Insert happens while typing, so any and all perf benefit is worth taking, and a convenient type already exists.

If `InsertTextFormat` is an enum, then it will be natively supported by MessagePach already.

Reply:
I tried TextChange but didn't like the need to convert from (line, character) to index text coordinates (just to send things over the wire). I "deconstructed" for now into LinePositionSpan (property annotated) and Text (string), will create maybe a custom wrapper if we feel we need it (though it doesn't look too bad is as I think)